internal/controller/notificationservice_controller.go

```diff
@@ -60,12 +60,28 @@ func (r *NotificationServiceReconciler) Reconcile(ctx context.Context, req ctrl.
 	}
 
 	logger.Info("Reconciling PipelineRun", "Name", pipelineRun.Name)
-	if !IsPipelineRunEnded(pipelineRun) &&
-		!IsFinalizerExistInPipelineRun(pipelineRun, NotificationPipelineRunFinalizer) {
-		err = AddFinalizerToPipelineRun(ctx, pipelineRun, r, NotificationPipelineRunFinalizer)
-		if err != nil {
-			logger.Error(err, "Failed to add finalizer", "pipelineRun", pipelineRun.Name)
-			return ctrl.Result{}, err
+
+	if !IsPipelineRunEnded(pipelineRun) {
+		// If pipeline has been deleted and has a finalizer
+		// Remove finalizer to allow deletion of the pipeline
+		if pipelineRun.GetDeletionTimestamp() != nil &&
+			IsFinalizerExistInPipelineRun(pipelineRun, NotificationPipelineRunFinalizer) {
```

Comment:
Would anything work differently (other than error message on the logs) if you did not check that the finalizer exist?

It could have made the code cleaner and also avoid possible race condition in which the finalizer is added/removed between the time of check and the time you add/remove it.

Reply:
As discussed, I removed the check for the finalizer